### PR TITLE
[scons] Enable link time optimization for ARM.

### PR DESCRIPTION
**Pull request by salkinium**
Does what the title says.

### CODE REVIEW

**Comment by ekiwi** [comment]
This looks promising.
I am currently running `scons check` against the `arm-none-eabi-gcc` that comes with Fedora 23. I will report the results as soon as it finishes.

**Comment by ekiwi** [comment]
@salkinium do you have time to test this on real hardware?
We have to make sure, that none of our startup code is removed and that the interrupt handlers still work as intended.

File: src/xpcc/architecture/platform/driver/core/cortex/syscalls.cpp.in [hunk 1]
```diff
@@ -290,11 +291,13 @@ void *malloc(size_t size)
 	return p;
 }
 
+__attribute__((used))
 void *realloc(void *p, size_t size)
 {
 	return tlsf_realloc(get_tlsf_for_ptr(p), p, size);
 }
 
+__attribute__((used))
```
Comment (ekiwi) [inline src/xpcc/architecture/platform/driver/core/cortex/syscalls.cpp.in:300]: what if the application does not use dynamic memory management?

Comment (salkinium) [inline src/xpcc/architecture/platform/driver/core/cortex/syscalls.cpp.in:300]: Yeah, this is the brute-force method to get something working.
And indeed, blinky links in the TLSF code, even though it does not use it.
Hmpf.

**Comment by salkinium** [comment]
Yes, I'll test on the Niklas-as-a-Service test farm.

**Comment by ekiwi** [comment]
OK, so the results are in and quite some stuff is failing.
As you are already in progress of fixing stuff, please just let me know, when I should run `scons check` again.

**Comment by salkinium** [comment]
Weird assembly failure here? https://travis-ci.org/roboterclubaachen/xpcc/jobs/128720517#L1012-L1019

**Comment by salkinium** [comment]
I'm not convinced that LTO can simply be switched on, without us taking a very detailed look at what it is throwing out.

**Comment by salkinium** [comment]
Enabling LTO will screw with linker sections, and it's [not going to get fixed](https://gcc.gnu.org/bugzilla/show_bug.cgi?id=65252).
I found out about this from [this LLVM talk](https://www.youtube.com/watch?v=hhaPAKUt35E).

**Pull request closed without merging**